### PR TITLE
GATE-1502 Update iomt-rubocop gem

### PR DESCRIPTION
**Pull request by ipetriniITH**

### CODE REVIEW

File: lib/iomt-rubocop/check_parentheses.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    # Define the IomtRubocop namespace.
+    module IomtRubocop
+      # Define cop for finding method calls that shouldn't be parenthesized.
+      class CheckParentheses < Cop
+        # Define message for cop.
+        MSG = 'Do not use parentheses for method calls that have a keyword status'.freeze
+
+        # Check methods for unnecessary parentheses.
+        def on_send(node)
+          return unless node.arguments?
```
Comment (jeremykirc) [inline lib/iomt-rubocop/check_parentheses.rb:14]: add new line after this line

File: lib/iomt-rubocop/check_parentheses.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          return unless node.parenthesized? && included_method?(node.method_name)
```
Comment (jeremykirc) [inline lib/iomt-rubocop/check_parentheses.rb:15]: combine these two with an ||

File: lib/iomt-rubocop/request_constant_comment.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        # Define regular expression.
+        COMMENT_STRUCTURE = /^# [A-Z].*[^.,?!;](\.){1}$/.freeze
+
+        # Check constants for comment unless the previous node is also a constant.
+        def on_casgn(node)
+          unless node.parent.nil? || node.sibling_index.zero?
+            previous_node = node.parent.children[node.sibling_index - 1]
+          end
+          return if previous_node.class.to_s == 'Symbol'
+          return unless previous_node.nil? || !previous_node.casgn_type?
```
Comment (jeremykirc) [inline lib/iomt-rubocop/request_constant_comment.rb:23]: can you combine this?

File: lib/iomt-rubocop/timestamps_blank_line.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      class TimestampsBlankLine < Cop
+        # Define message for cop.
+        MSG = 't.timestamps should have a preceding blank line'.freeze
+
+        # Define the matcher for the node.
+        def_node_matcher :timestamp_definition?, '(send (lvar :t) :timestamps)'
+
+        # Check node for a preceding blank line.
+        def on_send(node)
+          return unless timestamp_definition?(node)
```
Comment (jeremykirc) [inline lib/iomt-rubocop/timestamps_blank_line.rb:17]: add new line

File: default.yml [hunk 5]
```diff
@@ -24,6 +29,29 @@ AllCops:
...
+IomtRubocop/CheckParentheses:
+  IncludedMethods:
+    - GET
+    - POST
+    - PUT
+    - PATCH
+    - DELETE
+    - redirect_to
+    - render
+    - return
```
Comment (bwang-ith) [inline default.yml:41]: Why are the HTTP methods in all caps? When you call the methods, they are lower case.

sorry I never got you this list. Good start! Add the following.
- assert, assert_not
- callbacks like after_action, before_action, before_destroy, etc.
- associations like belongs_to, has_many, etc.
- validates
- protect_from_forgery
- raise
- resources
- migration methods like create_table

File: lib/iomt-rubocop/check_parentheses.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# Define namespace of Rubocop.
+module RuboCop
+  # Define namespace of Rubocop Cop.
+  module Cop
+    # Define the IomtRubocop namespace.
+    module IomtRubocop
+      # Define cop for finding method calls that shouldn't be parenthesized.
+      class CheckParentheses < Cop
```
Comment (bwang-ith) [inline lib/iomt-rubocop/check_parentheses.rb:8]: This cop needs to do _both_ the positive and negative check
- keyword methods must _not_ have parentheses, but also...
- all other methods _must_ have parentheses

File: lib/iomt-rubocop/request_constant_comment.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        # Define regular expression.
+        COMMENT_STRUCTURE = /^# [A-Z].*[^.,?!;](\.){1}$/.freeze
+
+        # Check constants for comment unless the previous node is also a constant.
+        def on_casgn(node)
+          unless node.parent.nil? || node.sibling_index.zero?
+            previous_node = node.parent.children[node.sibling_index - 1]
+          end
+          return if previous_node.class.to_s == 'Symbol' ||
+                    !previous_node.nil? &&
```
Comment (bwang-ith) [inline lib/iomt-rubocop/request_constant_comment.rb:23]: prefer not to use `!` when possible. Consider `.present?`

File: lib/iomt-rubocop/rspec_controller_format.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        def_node_matcher :it_definition?, '(send nil? :it (str _))'
+
+        # Define regular expressions.
+        DESCRIBE_REGEX = /'(GET|PATCH|POST|PUT|DELETE) .+'/.freeze
+        CONTEXT_REGEX = /'(when|with) .+'/.freeze
+        TEST_REGEX = /'should .+'/.freeze
+
+        # Check for every part of the test to be in the correct format.
+        def on_send(node)
+          describe_format(node) if describe_definition?(node)
```
Comment (bwang-ith) [inline lib/iomt-rubocop/rspec_controller_format.rb:21]: change these method names to read like booleans. e.g. `describe_definition?` -> `describe_definition_valid?`

File: lib/iomt-rubocop/rspec_controller_format.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+          test_format(node) if it_definition?(node)
+        end
+
+        private
+
+        # Check describe format.
+        def describe_format(node)
+          return if node.arguments.first.source =~ DESCRIBE_REGEX
+
+          msg = 'Wrong describe format. The format should follow: '\
```
Comment (bwang-ith) [inline lib/iomt-rubocop/rspec_controller_format.rb:32]: You're right that we do prefer single quotes, but if there are single quotes in the string, just use outer double quotes. That way you can avoid the escapes. More readable.

File: lib/iomt-rubocop/rspec_controller_format.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+                'describe \'<HTTP verb> <action>\' do'
+          add_offense(node, message: msg)
+        end
+
+        # Check context format.
+        def context_format(node)
+          return if node.arguments.first.source =~ CONTEXT_REGEX
+
+          msg = 'Wrong context format. The format should follow: '\
```
Comment (bwang-ith) [inline lib/iomt-rubocop/rspec_controller_format.rb:41]: same

File: lib/iomt-rubocop/rspec_controller_format.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                'context \'when|with <conditions>\' do'
+          add_offense(node, message: msg)
+        end
+
+        # Check test format.
+        def test_format(node)
+          return if node.arguments.first.source =~ TEST_REGEX
+
+          msg = 'Wrong context format. The format should follow: '\
```
Comment (bwang-ith) [inline lib/iomt-rubocop/rspec_controller_format.rb:50]: same for all

File: lib/iomt-rubocop/rspec_model_format.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
+# Define namespace of Rubocop.
+module RuboCop
+  # Define namespace of Rubocop Cop.
+  module Cop
+    # Define the IomtRubocop namespace.
+    module IomtRubocop
+      # Define cop for finding wrong formatted rspec model tests.
+      class RspecModelFormat < Cop
+        # Define the matcher for the node.
+        def_node_matcher :describe_definition?, '(send nil? :describe (str _))'
```
Comment (bwang-ith) [inline lib/iomt-rubocop/rspec_model_format.rb:10]: Again, make these method names read like boolean methods.

File: lib/iomt-rubocop/request_constant_comment.rb (hunk 7 above)
Comment (ipetriniITH) [reply lib/iomt-rubocop/request_constant_comment.rb:23]: Yeah I tried that but AST nodes wouldn't let me use `.present?`

File: lib/iomt-rubocop/rspec_controller_format.rb (hunk 8 above)
Comment (ipetriniITH) [reply lib/iomt-rubocop/rspec_controller_format.rb:21]: Actually, that def_node_matcher checks if a node is a describe definition, but not that it is valid, the validation comes with the REGEX, so maybe `is_describe_definition?` would be more clear, right?

File: default.yml [hunk 13]
```diff
@@ -24,6 +29,29 @@ AllCops:
...
+IomtRubocop/CheckParentheses:
+  IncludedMethods: &methods [ruby, get, post, patch, delete, redirect_to, render, return,
+                    belongs_to, has_one, has_many, has_and_belongs_to_many, puts, exit
+                    validates, protect_from_forgery, raise, resources, create_table,
+                    change_table, drop_table, add_column, add_index, boolean, integer,
+                    add_foreign_key, change_column, rename_column, rename_index, exec,
+                    rename_table, remove_column, remove_foreign_key, remove_index, set,
+                    validate, enum, include, require_relative, run, source, gem, error,
+                    require, can, head, index, string, datetime, bigint, json, to, not_to]
+  IncludedPatterns: &patterns [^assert, ^(before|after|around)_]
```
Comment (bwang-ith) [inline default.yml:41]: Nice.

File: lib/iomt-rubocop/check_parentheses.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        # Define message for cop.
+        MSG = 'Do not use parentheses for method calls'\
+                                   ' that have a keyword status'.freeze
```
Comment (bwang-ith) [inline lib/iomt-rubocop/check_parentheses.rb:11]: fix indent. Also, we put the space at the end of the upper line, not beginning of lower line.

File: lib/iomt-rubocop/request_constant_comment.rb (hunk 7 above)
Comment (bwang-ith) [reply lib/iomt-rubocop/request_constant_comment.rb:23]: fair enough, you're good then

**Pull request merged**